### ISSUE TITLE
Identifying metagenomic contigs matching reference genomes with gather?

### ISSUE THREAD

**Issue by jcmcnch** [dib-lab/sourmash#528]
Hi Titus and dib-lab,

I have an assembly of ~2.5 million contigs, and I'd like to use sourmash to search each contig against a small database of reference genomes. My plan is to then take matching contigs to try and obtain better bins for the organisms matching these reference genomes (which are always highly fragmented and so do not get well binned into MAGs). I thought sourmash should be a quick way of doing this.

Here's my workflow so far:

1. Create a reference database for reference genomes and signatures for metagenomic contigs, e.g.:

Ref genomes:
sourmash compute -k 21,31,51 --scaled=1000 ../cultivar_genomes/*fna
sourmash index -k 21 cultivars_sourmash_db *.sig

Contigs:
sourmash compute -k 21,31,51 --scaled=1000 --singleton contigs.fa

2. Find which contigs match genomes using gather:
sourmash gather --threshold-bp 500 --ksize 21 contigs.fa.sig cultivars_sourmash_db.sig
*My minimum contig size is 500 bp which is why I chose the low threshold

However, when I tried searching the signature derived from the command for generating the contig signatures (sourmash compute -k 21,31,51 --scaled=1000 --singleton contigs.fa) against the database of reference genomes, it didn't work properly, giving a generic error about sequence type (can get the specific error if it would be helpful). I have since tried to split the contig file manually into individual fasta files and then rerun the signature creation step with a bash loop. But this was apparently a bad idea because it's much slower than the --singleton method (which completed the calculation in ~3 hours whereas the loop is still only about 1/7 done after running overnight).

Do you have any suggestions for how to carry out this analysis? Any advice you can offer will be greatly appreciated!

Cheers,

Jesse

**Comment by ctb** [comment]
Hi Jesse,

check out --

http://ivory.idyll.org/blog/2018-detecting-contamination-in-long-read-assemblies.html

for some ideas. Seem interesting?

I'm currently traveling, so it may take me a few days to get back to you, but
a few quick thoughts.

> However, when I tried searching the signature derived from the command for generating the contig signatures (sourmash compute -k 21,31,51 --scaled=1000 --singleton contigs.fa) against the database of reference genomes, it didn't work properly, giving a generic error about sequence type (can get the specific error if it would be helpful).

yes, please send error! I don't see anything obviously problematic above, so...

> I have since tried to split the contig file manually into individual fasta files and then rerun the signature creation step with a bash loop. But this was apparently a bad idea because it's much slower than the --singleton method (which completed the calculation in ~3 hours whereas the loop is still only about 1/7 done after running overnight).

yeah... the database search isn't super fast :(.

One thing you could do is this:

1. run sourmash gather contigs.sig db.sig --save-matches contigs.matches.sig
2. then do your individual individual fasta record search against contigs.matches.sig

this is a silly optimization that we could/should simply bake in to the process
somehow but is not. basically you're just whittling the database size down
to those that matter for each collection of contigs.

more in a bit.

best,
--titus
-- 
C. Titus Brown, ctbrown@ucdavis.edu

**Comment by jcmcnch** [comment]
Hi Titus,

Thanks very much for your prompt reply!

Here is the error I get when I try to run the signature generated from this command:

sourmash compute -k 21,31,51 --scaled=1000 ../cultivar_genomes/*fna

against the database constructed from references genomes:

sourmash gather --threshold-bp 500 --ksize 21 test.sig cultivars_sourmash_db.sbt.json

error:

...sig loading 9
When loading query from "test.sig"
10 signatures matching ksize and molecule type;
need exactly one. Specify --ksize or --dna/--protein.

I have specified the --ksize in the command, and I tried it with or without the --dna flag present but doesn't work either way. Seems like gather can't interpret a multiple signature file? Is this the expected behaviour?

Thanks also for the blog post, it does look interesting! I wanted to ask a quick clarification related to Fred's comments on the blog post at the bottom. You guys were discussing a 2 kB limit for contig size because of the number of samples of kmers taken by sourmash from sequences. If I want to get matches to shorter contigs (i.e. 500 bp), could I do this by altering the --scaled parameter when constructing the signatures? Perhaps to something like --scaled=100?

Thanks very much again for your help and safe travels,

Jesse
